### DISCUSSION TITLE
Compose file cleanup

### DISCUSSION THREAD

**Discussion by schongallam** [narfindustries/digiheals-public#13] [Ideas]
Per the 6/5 meeting, I'm cleaning up the Compose file a bit.  Initially, I propose:
- specifying docker image tags more consistently, for better reproducibility & version control
- re-using docker images where feasible
  - `mysql` and `openmrs-db` services can derive from a common, updated mariadb image
  - `openemr` and `gnuhealth` can derive from a common alpine image (in their respective Dockerfiles)
- adding healthchecks
- adding missing `depends_on`
- standardize the order of YAML keys within each service (for style)

Also, what would people think about consolidating the `public` and `private` docker networks?
- Upside: simplicity.
- Downside: I don't see any.
- `private` seems to only apply as a holdover in `postgres`, anyway

I haven't poked around with all the services yet, but I'm sure I'll find more stuff to optimize, later.
Any objections/concerns/additions/etc. on these goals so far?

Thanks for your thoughts,
Malcolm

**Comment by schongallam** [comment]
Another thought - some of the service names for databases (for openemr, gnuhealth, openmrs) seem to be holdovers.  What do folks think about renaming them to something more informative, like db-gnuhealth, db-openemr, db-openmrs to correspond with the services they support?  (I want to make sure this won't break anybody's workflows)

**Comment by prashantbarca** [comment]
Yes, I agree with all the points here pretty much. Health checks, depends_on keys make sense. The private and public networks are probably not very useful in our context---as long as they are all accessible via the docker compose network it should be fine. 

How do you propose specifying the image tags more consistently? 

I wonder if there is a static analysis tool we can use to standardize the order of the YAML keys.

**Reply by schongallam** [reply]
Currently, images subject to unexpected change include postgres, gnuhealth, openmrs, ibm-fhir, hapi, blaze, synthea, and openemr (which even has a misleading version comment).  Several have had github activity and even minor version updates within the last few weeks.  I think a breaking change would be unlikely, but preventing that is low hanging fruit.

(fhir-vista is also `latest`, but it's pretty stale, I doubt anyone's expecting a surprise update.)

To specify image tags more consistently, I would just pull specific tags (for images on hub) instead of just the default / `latest`.

And for images that are built locally, I would make explicit which version tarballs are downloaded, and tag them in the compose file accordingly. i.e.,

gnuhealth/Dockerfile:
```
ENV GNUHEALTH_VERSION=4.4.0
...
RUN wget https://ftp.gnu.org/gnu/health/gnuhealth-${GNUHEALTH_VERSION}.tar.gz
```

docker-compose.yml:
```
    build: ./ehrgarden/gnuhealth
    image: gnuhealth:4.4.0
````
(or similar)

For specific versions, alpine3.20 seems to work fine in place of other alpine versions.
MariaDB 11.4.2 also works for openemr and openmrs just fine (with some syntax updates).

Having done that for all images, my `docker images` output is more informative. I don't have anything tagged 'latest' any more.  I don't have to worry about surprise breaking changes.

**Comment by schongallam** [comment]
For key order, I haven't found any definitive style guides or static analysis-based recommendations.  Mine is in this order, but I don't have strong feelings one way or another:
```
depends_on
build
image
command
restart
environment
ports
volumes
healthcheck
```

**Comment by prashantbarca** [comment]
I think the items you have proposed are reasonable. Would you be able to submit a PR for this?

**Reply by schongallam** [reply]
Most definitely. I am almost ready to submit-- just a few more tweaks and validations to run.  (Probably after the weekend)

**Reply by schongallam** [reply]
Submitted!  There may be some more I could do later, but I'm going to be pretty busy the rest of the week.  I wanted to submit what I had before I got too far behind on potential merge conflicts (shouldn't be any for this PR).
